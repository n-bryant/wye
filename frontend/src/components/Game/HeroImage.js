import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";

import styles from "./HeroImage.styles";

export const PLACEHOLDER_PATH = "/steamHeroPlaceholder.jpg";

/**
 * renders a hero image with a placeholder src value
 */
export const HeroImage = props => {
  const classnames = HeroImage.classnames(props);
  const { altText, imageSrc, squaredBottom } = props;
  const [hasBrokenSrc, setHasBrokenSrc] = React.useState(false);
  const [loaded, setLoaded] = React.useState(false);

  const onLoad = () => {
    setLoaded(true);
  };
  const onError = () => {
    setHasBrokenSrc(true);
  };

  return (
    <div className={classnames.root()}>
      <img
        className={classnames.element("image", {
          hidden: !loaded,
          squaredBottom
        })}
        onError={onError}
        onLoad={onLoad}
        src={hasBrokenSrc ? PLACEHOLDER_PATH : imageSrc}
        alt={altText}
      />
      <img
        className={classnames.element("placeholder", {
          hidden: loaded
        })}
        src={PLACEHOLDER_PATH}
        alt={altText}
      />
    </div>
  );
};
HeroImage.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}HeroImage`
);
HeroImage.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    image: PropTypes.string,
    imageSquaredBottom: PropTypes.string,
    imageHidden: PropTypes.string,
    placeholder: PropTypes.string,
    placeholderHidden: PropTypes.string
  }),
  // alt text for the image
  altText: PropTypes.string,
  // the location of the image to render
  imageSrc: PropTypes.string,
  // whether the bottom of the image should have a squared bottom
  squaredBottom: PropTypes.bool
};
HeroImage.defaultProps = {
  classes: {},
  altText: "",
  imageSrc: ""
};

export default withStyles(styles)(HeroImage);
