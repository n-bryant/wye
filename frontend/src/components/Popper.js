import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";

import styles from "./Popper.styles";

export const constants = {
  POPPER_WIDTH: 300,
  POPPER_HEIGHT: 354,
  HEADER_HEIGHT: 56,
  POPPER_MARGIN: 12,
  ARROW_ORIENTATION: {
    LEFT: "left",
    RIGHT: "right"
  }
};

/**
 * determine the properties to apply to the popper
 * - based on the anchor's dimensions/positioning
 * @param {Object} anchorRect
 * @returns {Object}
 */
export const getPopperProps = anchorRect => {
  const {
    POPPER_WIDTH,
    POPPER_MARGIN,
    HEADER_HEIGHT,
    POPPER_HEIGHT,
    ARROW_ORIENTATION
  } = constants;

  // check to make sure if/where the popper has the requisite space to render
  const hasRoomRight =
    anchorRect.right + POPPER_WIDTH + POPPER_MARGIN < window.innerWidth;
  const hasRoomLeft = anchorRect.left > POPPER_WIDTH + POPPER_MARGIN;
  const hasRoomBottom = window.innerHeight - anchorRect.top > POPPER_HEIGHT;
  const hasRoomTop = window.innerHeight - HEADER_HEIGHT > POPPER_HEIGHT;

  // we don't want to render if there isn't space available
  let displayProperties = {
    style: {
      display: "none"
    }
  };
  if (!hasRoomTop) {
    return displayProperties;
  }

  // set initial arrow position
  let arrowStyle = {
    top: `calc(${anchorRect.height}px / 4)`
  };

  // set the left/right orientation of the popper relative to the anchor and based on available space
  if (hasRoomRight) {
    displayProperties = {
      arrowOrientation: ARROW_ORIENTATION.LEFT,
      arrowStyle,
      style: {
        left: `${anchorRect.width + POPPER_MARGIN}px`
      }
    };
  } else if (hasRoomLeft) {
    displayProperties = {
      arrowOrientation: ARROW_ORIENTATION.RIGHT,
      arrowStyle,
      style: {
        left: `-${POPPER_WIDTH + POPPER_MARGIN}px`
      }
    };
  }

  // calculate and apply vertical offset if there is not enough room below
  // the anchor to render the popper
  if (hasRoomBottom === false && displayProperties) {
    const offset =
      POPPER_HEIGHT -
      (window.innerHeight - anchorRect.bottom + anchorRect.height);
    displayProperties = {
      ...displayProperties,
      arrowStyle: {
        // make sure the arrow position is also adjusted
        top: `calc(${anchorRect.height}px / 4 + ${offset}px)`
      },
      style: {
        ...displayProperties.style,
        top: -offset
      }
    };
  }

  return displayProperties;
};

/**
 * renders a Popper with the provided content
 * - positioning based on screen dimensions and anchor positioning
 * - options for displaying arrow
 */
export const Popper = props => {
  const classnames = Popper.classnames(props);
  const { anchorEl, displayArrow } = props;
  const { POPPER_WIDTH, POPPER_HEIGHT, ARROW_ORIENTATION } = constants;
  const [displayProperties] = React.useState(
    getPopperProps(anchorEl.getBoundingClientRect())
  );

  return (
    <div
      className={classnames.element("root")}
      style={{
        ...displayProperties.style,
        width: POPPER_WIDTH,
        height: POPPER_HEIGHT
      }}
    >
      {displayArrow && (
        <div
          style={displayProperties.arrowStyle}
          className={classnames.element("arrow", {
            left: displayProperties.arrowOrientation === ARROW_ORIENTATION.LEFT,
            right:
              displayProperties.arrowOrientation === ARROW_ORIENTATION.RIGHT
          })}
        />
      )}
      <div className={classnames.element("content")}>{props.children}</div>
    </div>
  );
};
Popper.classnames = createClassNameHelper(`${JSS_CLASS_NAME_PREFIX}Popper`);
Popper.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    content: PropTypes.string,
    arrow: PropTypes.string,
    arrowLeft: PropTypes.string,
    arrowRight: PropTypes.string
  }),
  anchorEl: PropTypes.object.isRequired,
  displayArrow: PropTypes.bool
};
Popper.defaultProps = {
  classes: {},
  displayArrow: true
};

export default withStyles(styles)(Popper);
