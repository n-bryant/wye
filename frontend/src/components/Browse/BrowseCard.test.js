import { act } from "react-dom/test-utils";
import { MockedProvider } from "@apollo/react-testing";

import Card from "@material-ui/core/Card";
import Chip from "@material-ui/core/Chip";
import CardActions from "@material-ui/core/CardActions";

import HeroImage from "../Game/HeroImage";
import Popper from "../Popper";
import HighlightTrailerQuery, {
  BrowseCard,
  VARIANTS_MAP,
  GET_HIGHLIGHT_TRAILER
} from "./BrowseCard";

describe("BrowseCard", () => {
  const props = {
    classes: {
      root: "root",
      header: "header",
      sm: "sm",
      md: "md",
      lg: "lg",
      lib: "lib",
      maxSize: "maxSize",
      media: "media",
      gameLinkHeader: "gameLinkHeader",
      gameLinkSm: "gameLinkSm",
      gameLinkMd: "gameLinkMd",
      gameLinkLg: "gameLinkLg",
      gameLinkLib: "gameLinkLib",
      gameLinkMaxSize: "gameLinkMaxSize",
      content: "content",
      title: "title",
      category: "category",
      genresContainer: "genresContainer",
      chip: "chip",
      chipLabel: "chipLabel",
      trailer: "trailer",
      actionArea: "actionArea",
      priceWidget: "priceWidget",
      gameLink: "gameLink",
      actionsContainer: "actionsContainer",
      actionLink: "actionLink",
      actionLinkLabel: "actionLinkLabel",
      horizontal: "horizontal",
      gameLinkHorizontal: "gameLinkHorizontal",
      gameLinkHorizontalWide: "gameLinkHorizontalWide",
      horizontalContent: "horizontalContent",
      horizontalContentWide: "horizontalContentWide",
      contentWrapper: "contentWrapper",
      horizontalTitle: "horizontalTitle",
      horizontalSubContent: "horizontalSubContent",
      usersContainer: "usersContainer"
    },
    data: {
      appid: "1",
      name: "foo",
      publishers: ["a", "b"],
      developers: ["a", "b"],
      userRating: 93,
      ownersFormatted: "10,000 .. 20,000",
      genres: ["a", "b", "c", "d", "e"],
      freeToPlay: true,
      onSale: false,
      discount: 0,
      initialPrice: 0,
      finalPrice: 0,
      headerImage: "header",
      capsuleSm: "small capsule",
      capsuleMd: "md capsule",
      capsuleLg: "lg capsule",
      libraryCapsule: "lib capsule"
    },
    variant: "header",
    trailerPath: "trailer"
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<BrowseCard {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should apply variant styling based on the variant prop", () => {
    Object.keys(VARIANTS_MAP).forEach(variant => {
      wrapper.setProps({
        variant
      });
      wrapper.update();
      expect(wrapper.find(Card).hasClass(variant)).toBeTruthy();
      const heroImage = wrapper.find(HeroImage);
      expect(
        heroImage
          .props()
          .className.toLowerCase()
          .indexOf(`media${variant}`)
      ).toBeTruthy();
    });
  });

  it("should render a horizontal CardContent if the variant is horizontal", () => {
    wrapper.setProps({
      variant: "horizontal",
      userDetails: [
        {
          id: "1",
          profileUrl: "url",
          avatarName: "name",
          avatarImageUrl: "img",
          onlineStatus: "Online"
        }
      ],
      playtimeAndOwnershipData: {
        ownedBy: ["1"],
        recentlyPlayedBy: ["1"],
        playtime: [
          {
            steamId: "1",
            hoursPlayed: 3.5
          }
        ]
      }
    });
    wrapper.update();
    const horizontalContent = wrapper.findWhere(n =>
      n.hasClass(props.classes.horizontalContent)
    );
    expect(horizontalContent.length).toBe(1);
    expect(toJson(horizontalContent)).toMatchSnapshot();
  });

  it("should set the image src for the HeroImage based on the variant", () => {
    Object.keys(VARIANTS_MAP).forEach(variant => {
      wrapper.setProps({
        variant
      });
      wrapper.update();
      const heroImage = wrapper.find(HeroImage);
      expect(heroImage.prop("imageSrc")).toBe(
        props.data[VARIANTS_MAP[variant].image]
      );
    });
  });

  it("should render CardActions if the cardActionLabel, cardActionLinkPath, and cardActionHref props are defined", () => {
    wrapper.setProps({
      cardActionLabel: "label",
      cardActionLinkPath: "path",
      cardActionHref: "href"
    });
    wrapper.update();
    const cardActions = wrapper.find(CardActions);
    expect(cardActions.length).toBe(1);
    expect(toJson(cardActions)).toMatchSnapshot();
  });

  describe("popper content", () => {
    let wrapperWithState;
    beforeEach(() => {
      const anchorEl = {};
      const setAnchorEl = jest.fn();
      const useStateSpy = jest.spyOn(React, "useState");
      useStateSpy.mockImplementation(() => [anchorEl, setAnchorEl]);
      wrapperWithState = shallow(<BrowseCard {...props} />);
    });

    it("should only render a Popper if an anchorEl is defined", () => {
      expect(wrapper.find(Popper).length).toBe(0);

      expect(wrapperWithState.find(Popper).length).toBe(1);
      expect(wrapperWithState.find(Popper)).toMatchSnapshot();
    });

    it("should render the headerImage if the trailerPath is not available", () => {
      expect(wrapperWithState.find("video").length).toBe(1);
      expect(wrapperWithState.find(Popper).find("img").length).toBe(0);

      wrapperWithState.setProps({
        trailerPath: ""
      });
      wrapperWithState.update();
      expect(wrapperWithState.find("video").length).toBe(0);
      const trailerSubstitute = wrapperWithState.find(Popper).find("img");
      expect(trailerSubstitute.prop("src")).toBe(props.data.headerImage);
    });

    it("should only render the first three genres available", () => {
      expect(wrapperWithState.find(Chip).length).toBe(3);
    });
  });
});

describe("HighlightTrailerQuery", () => {
  const props = {
    data: {
      appid: "foo"
    }
  };

  it("should render successfully", async () => {
    const mocks = [
      {
        request: {
          query: GET_HIGHLIGHT_TRAILER,
          variables: {
            gameId: "1"
          }
        },
        result: {
          data: {
            getHighlightTrailer: props.data.appid
          }
        }
      }
    ];
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <MockedProvider mocks={mocks} addTypename={false}>
          <HighlightTrailerQuery {...props} />
        </MockedProvider>
      );
    });
    expect(toJson(wrapper.find(HighlightTrailerQuery))).toMatchSnapshot();
  });
});
