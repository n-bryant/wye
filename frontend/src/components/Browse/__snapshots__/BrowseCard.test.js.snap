// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowseCard popper content should only render a Popper if an anchorEl is defined 1`] = `
<WithStyles(Popper)
  anchorEl={Object {}}
>
  <WithStyles(ForwardRef(Typography))
    className="wye-BrowseCard__title title"
    variant="h3"
  >
    foo
  </WithStyles(ForwardRef(Typography))>
  <video
    autoPlay={true}
    className="wye-BrowseCard__trailer trailer"
    height="215"
    loop={true}
    muted={true}
    playsInline={true}
    poster="/featuredItemSpacer.png"
    preload="none"
    src="trailer#t=0"
    width="384"
  >
    foo
  </video>
  <WithStyles(ForwardRef(Typography))
    className="wye-BrowseCard__withEllipsis withEllipsis"
    variant="body2"
  >
    <span
      className="wye-BrowseCard__category category"
    >
      Developers: 
    </span>
     
    a, b
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    className="wye-BrowseCard__withEllipsis withEllipsis"
    gutterBottom={true}
    variant="body2"
  >
    <span
      className="wye-BrowseCard__category category"
    >
      Publishers: 
    </span>
     
    a, b
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    variant="body2"
  >
    <span
      className="wye-BrowseCard__category category"
    >
      User Rating:
       
    </span>
     
    93
    %
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    gutterBottom={true}
    variant="body2"
  >
    <span
      className="wye-BrowseCard__category category"
    >
      Owners: 
    </span>
     
    10,000 .. 20,000
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    variant="body2"
  >
    Genres:
  </WithStyles(ForwardRef(Typography))>
  <div
    className="wye-BrowseCard__genresContainer genresContainer"
  >
    <WithStyles(ForwardRef(Chip))
      className="wye-BrowseCard__chip chip"
      classes={
        Object {
          "label": "wye-BrowseCard__chipLabel chipLabel",
        }
      }
      key="0"
      label="a"
      size="small"
    />
    <WithStyles(ForwardRef(Chip))
      className="wye-BrowseCard__chip chip"
      classes={
        Object {
          "label": "wye-BrowseCard__chipLabel chipLabel",
        }
      }
      key="1"
      label="b"
      size="small"
    />
    <WithStyles(ForwardRef(Chip))
      className="wye-BrowseCard__chip chip"
      classes={
        Object {
          "label": "wye-BrowseCard__chipLabel chipLabel",
        }
      }
      key="2"
      label="c"
      size="small"
    />
  </div>
</WithStyles(Popper)>
`;

exports[`BrowseCard should render CardActions if the cardActionLabel, cardActionLinkPath, and cardActionHref props are defined 1`] = `
<WithStyles(ForwardRef(CardActions))
  className="wye-BrowseCard__actionsContainer actionsContainer"
>
  <Link
    as="path"
    href="href"
  >
    <a
      className="wye-BrowseCard__actionLink actionLink"
    >
      <WithStyles(ForwardRef(Typography))
        className="wye-BrowseCard__actionLinkLabel actionLinkLabel"
        variant="body1"
      >
        label
      </WithStyles(ForwardRef(Typography))>
    </a>
  </Link>
</WithStyles(ForwardRef(CardActions))>
`;

exports[`BrowseCard should render a horizontal CardContent if the variant is horizontal 1`] = `
<WithStyles(ForwardRef(CardContent))
  className="wye-BrowseCard__horizontalContent horizontalContent"
>
  <div
    className="wye-BrowseCard__usersContainer usersContainer"
  >
    <WithStyles(UserAvatar)
      data={
        Object {
          "avatarImageUrl": "img",
          "avatarName": "name",
          "id": "1",
          "onlineStatus": "Online",
          "profileUrl": "url",
        }
      }
      gameDetails={
        Object {
          "hoursPlayed": 3.5,
          "ownsGame": true,
          "recentlyPlayedGame": true,
        }
      }
      key="1"
    />
  </div>
  <div
    className="wye-BrowseCard__contentWrapper contentWrapper"
  >
    <WithStyles(ForwardRef(Typography))
      className="wye-BrowseCard__horizontalTitle horizontalTitle"
      variant="h2"
    >
      foo
    </WithStyles(ForwardRef(Typography))>
    <div
      className="wye-BrowseCard__horizontalSubContent horizontalSubContent"
    >
      <WithStyles(PriceWidget)
        data={
          Object {
            "discountPercentage": 0,
            "finalFormatted": "$0.00",
            "freeToPlay": true,
            "initialFormatted": "$0.00",
            "onSale": false,
          }
        }
      />
    </div>
  </div>
</WithStyles(ForwardRef(CardContent))>
`;

exports[`BrowseCard should render successfully 1`] = `
<WithStyles(ForwardRef(Card))
  className="wye-BrowseCard wye-BrowseCard--header root header"
>
  <Link
    as="/game/1"
    href="/game/[id]"
  >
    <a
      className="wye-BrowseCard__gameLink wye-BrowseCard__gameLink--header gameLink gameLinkHeader"
    >
      <WithStyles(ForwardRef(CardActionArea))
        className="wye-BrowseCard__actionArea actionArea"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <WithStyles(ForwardRef(CardContent))
          className="wye-BrowseCard__content content"
        >
          <WithStyles(HeroImage)
            altText="foo"
            className="wye-BrowseCard__media media"
            imageSrc="header"
            placeholderPath="header"
            squaredBottom={false}
            squaredRight={false}
          />
        </WithStyles(ForwardRef(CardContent))>
      </WithStyles(ForwardRef(CardActionArea))>
    </a>
  </Link>
  <div
    className="wye-BrowseCard__priceWidget priceWidget"
  >
    <WithStyles(PriceWidget)
      data={
        Object {
          "discountPercentage": 0,
          "finalFormatted": "$0.00",
          "freeToPlay": true,
          "initialFormatted": "$0.00",
          "onSale": false,
        }
      }
      skinny={true}
    />
  </div>
</WithStyles(ForwardRef(Card))>
`;
